Rebased. There's one test failure now:
```
make test PKG=./storage TESTS=TestReplicateAfterTruncation
go test -v  -i ./storage
github.com/cockroachdb/cockroach/storage
github.com/cockroachdb/cockroach/server
go test  -run TestReplicateAfterTruncation -timeout 1m10s ./storage
I160601 20:57:49.518827 rand.go:76  Random seed: 8201202425187532114
I160601 20:57:49.520806 storage/engine/rocksdb.go:137  opening in memory rocksdb instance
W160601 20:57:49.521765 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
W160601 20:57:49.523413 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160601 20:57:49.524428 storage/replica_command.go:1516  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 1s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=1970-01-01 00:00:00 +0000 UTC]
I160601 20:57:49.524686 storage/engine/rocksdb.go:137  opening in memory rocksdb instance
W160601 20:57:49.525308 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160601 20:57:49.526095 gossip/gossip.go:923  starting client to 127.0.0.1:62781
I160601 20:57:49.528369 storage/replica_command.go:1372  range 1: attempting to truncate raft logs for another range 0. Normally this is due to a merge and can be ignored.
I160601 20:57:49.528692 storage/replica_raftstorage.go:560  generated snapshot for range 1 at index 15 in 93.583µs. encoded size=2239, 26 KV pairs, 5 log entries
I160601 20:57:49.531021 storage/replica.go:1385  raft: proposing ADD_REPLICA {2 2 2} for range 1
I160601 20:57:49.531353 storage/replica_raftstorage.go:655  replica 2 received snapshot for range 1 at index 15. encoded size=2239, 26 KV pairs, 5 log entries
I160601 20:57:49.531471 storage/replica_raftstorage.go:657  replica 2 applied snapshot for range 1 in 96.095µs
--- FAIL: TestReplicateAfterTruncation (0.01s)
	client_raft_test.go:510: expected stats on new range:
		{LastUpdateNanos:0 IntentAge:0 GCBytesAge:0 LiveBytes:1522 LiveCount:20 KeyBytes:444 KeyCount:20 ValBytes:1096 ValCount:21 IntentBytes:0 IntentCount:0 SysBytes:312 SysCount:5}
		to equal old:
		{LastUpdateNanos:0 IntentAge:0 GCBytesAge:0 LiveBytes:1538 LiveCount:20 KeyBytes:468 KeyCount:20 ValBytes:1158 ValCount:23 IntentBytes:0 IntentCount:0 SysBytes:355 SysCount:5}
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	0.047s
make: *** [test] Error 1
```

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -20,7 +20,6 @@ import (
 	"bytes"
 	"fmt"
 	"reflect"
-	"strings"
 	"sync"
 	"sync/atomic"
 	"testing"
@@ -389,7 +388,6 @@ func TestFailedReplicaChange(t *testing.T) {
 				if et, ok := filterArgs.Req.(*roachpb.EndTransactionRequest); ok && et.Commit {
 					return roachpb.NewErrorWithTxn(util.Errorf("boom"), filterArgs.Hdr.Txn)
 				}
-				return nil
 			}
 			return nil
 		}
@@ -401,13 +399,13 @@ func TestFailedReplicaChange(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	err = rng.ChangeReplicas(roachpb.ADD_REPLICA,
+	if err := rng.ChangeReplicas(roachpb.ADD_REPLICA,
 		roachpb.ReplicaDescriptor{
 			NodeID:  mtc.stores[1].Ident.NodeID,
 			StoreID: mtc.stores[1].Ident.StoreID,
-		}, rng.Desc())
-	if err == nil || !strings.Contains(err.Error(), "boom") {
-		t.Fatalf("did not get expected error: %s", err)
+		}, rng.Desc(),
+	); !testutils.IsError(err, "boom") {
+		t.Fatalf("did not get expected error: %v", err)
 	}
 
 	// After the aborted transaction, r.Desc was not updated.
@@ -424,12 +422,12 @@ func TestFailedReplicaChange(t *testing.T) {
 	// are pushable by making the transaction abandoned.
 	mtc.manualClock.Increment(10 * base.DefaultHeartbeatInterval.Nanoseconds())
 
-	err = rng.ChangeReplicas(roachpb.ADD_REPLICA,
+	if err := rng.ChangeReplicas(roachpb.ADD_REPLICA,
 		roachpb.ReplicaDescriptor{
 			NodeID:  mtc.stores[1].Ident.NodeID,
 			StoreID: mtc.stores[1].Ident.StoreID,
-		}, rng.Desc())
-	if err != nil {
+		}, rng.Desc(),
+	); err != nil {
 		t.Fatal(err)
 	}
 
@@ -511,9 +509,13 @@ func TestReplicateAfterTruncation(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	if mvcc, mvcc2 := rng.GetMVCCStats(), rng2.GetMVCCStats(); mvcc2 != mvcc {
-		t.Fatalf("expected stats on new range:\n%+v\nto equal old:\n%+v", mvcc2, mvcc)
-	}
+
+	util.SucceedsSoon(t, func() error {
+		if mvcc, mvcc2 := rng.GetMVCCStats(), rng2.GetMVCCStats(); mvcc2 != mvcc {
+			return util.Errorf("expected stats on new range:\n%+v\nto equal old:\n%+v", mvcc2, mvcc)
+		}
+		return nil
+	})
 
 	// Send a third command to verify that the log states are synced up so the
 	// new node can accept new commands.
@@ -712,7 +714,8 @@ func TestChangeReplicasDescriptorInvariant(t *testing.T) {
 				NodeID:  mtc.stores[storeNum].Ident.NodeID,
 				StoreID: mtc.stores[storeNum].Ident.StoreID,
 			},
-			desc)
+			desc,
+		)
 	}
 
 	// Retain the descriptor for the range at this point.

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -693,12 +693,12 @@ func (m *multiTestContext) replicateRange(rangeID roachpb.RangeID, dests ...int)
 			m.t.Fatal(err)
 		}
 
-		err = rep.ChangeReplicas(roachpb.ADD_REPLICA,
+		if err := rep.ChangeReplicas(roachpb.ADD_REPLICA,
 			roachpb.ReplicaDescriptor{
 				NodeID:  m.stores[dest].Ident.NodeID,
 				StoreID: m.stores[dest].Ident.StoreID,
-			}, &desc)
-		if err != nil {
+			}, &desc,
+		); err != nil {
 			m.t.Fatal(err)
 		}
 	}
@@ -708,8 +708,8 @@ func (m *multiTestContext) replicateRange(rangeID roachpb.RangeID, dests ...int)
 		for _, dest := range dests {
 			// Use LookupRange(keys) instead of GetRange(rangeID) to ensure that the
 			// snapshot has been transferred and the descriptor initialized.
-			if m.stores[dest].LookupReplica(startKey, nil) == nil {
-				return util.Errorf("range not found on store %d", dest)
+			if store := m.stores[dest]; store.LookupReplica(startKey, nil) == nil {
+				return util.Errorf("range %d not found on store %+v", rangeID, store.Ident)
 			}
 		}
 		return nil
@@ -733,12 +733,12 @@ func (m *multiTestContext) unreplicateRange(rangeID roachpb.RangeID, dest int) {
 		m.t.Fatal(err)
 	}
 
-	err = rep.ChangeReplicas(roachpb.REMOVE_REPLICA,
+	if err := rep.ChangeReplicas(roachpb.REMOVE_REPLICA,
 		roachpb.ReplicaDescriptor{
 			NodeID:  m.idents[dest].NodeID,
 			StoreID: m.idents[dest].StoreID,
-		}, &desc)
-	if err != nil {
+		}, &desc,
+	); err != nil {
 		m.t.Fatal(err)
 	}
 }

--- a/storage/helpers_test.go
+++ b/storage/helpers_test.go
@@ -22,15 +22,9 @@
 package storage
 
 import (
-	"time"
-
-	"github.com/coreos/etcd/raft"
-	"github.com/coreos/etcd/raft/raftpb"
-
 	"github.com/cockroachdb/cockroach/client"
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/storage/engine"
-	"github.com/cockroachdb/cockroach/util/retry"
 )
 
 // ComputeMVCCStats immediately computes correct total MVCC usage statistics
@@ -120,39 +114,4 @@ func (r *Replica) GetLastIndex() (uint64, error) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 	return r.LastIndex()
-}
-
-// GetSnapshot wraps Snapshot() but does not require the replica lock
-// to be held and it will block instead of returning
-// ErrSnapshotTemporaryUnavailable.
-func (r *Replica) GetSnapshot() (raftpb.Snapshot, error) {
-	retryOptions := retry.Options{
-		InitialBackoff: 1 * time.Millisecond,
-		MaxBackoff:     50 * time.Millisecond,
-		Multiplier:     2,
-	}
-	for retry := retry.Start(retryOptions); retry.Next(); {
-		r.mu.Lock()
-		snap, err := r.Snapshot()
-		snapshotChan := r.mu.snapshotChan
-		r.mu.Unlock()
-		if err == raft.ErrSnapshotTemporarilyUnavailable {
-			if snapshotChan == nil {
-				// The call to Snapshot() didn't start an async process due to
-				// rate limiting. Try again later.
-				continue
-			}
-			var ok bool
-			snap, ok = <-snapshotChan
-			if ok {
-				return snap, nil
-			}
-			// Each snapshot worker's output can only be consumed once.
-			// We could be racing with raft itself, so if we get a closed
-			// channel loop back and try again.
-		} else {
-			return snap, err
-		}
-	}
-	panic("unreachable") // due to infinite retries
 }

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -1382,7 +1382,7 @@ func defaultProposeRaftCommandLocked(r *Replica, p *pendingCmd) error {
 		if crt := etr.InternalCommitTrigger.GetChangeReplicasTrigger(); crt != nil {
 			// EndTransactionRequest with a ChangeReplicasTrigger is special because raft
 			// needs to understand it; it cannot simply be an opaque command.
-			log.Infof("raft: proposing %s %v for range %d", crt.ChangeType, crt.Replica, p.raftCmd.RangeID)
+			log.Infof("raft: proposing %s %+v for range %d", crt.ChangeType, crt.Replica, p.raftCmd.RangeID)
 
 			ctx := ConfChangeContext{
 				CommandID: string(p.idKey),
@@ -1394,12 +1394,11 @@ func defaultProposeRaftCommandLocked(r *Replica, p *pendingCmd) error {
 				return err
 			}
 
-			return r.mu.raftGroup.ProposeConfChange(
-				raftpb.ConfChange{
-					Type:    changeTypeInternalToRaft[crt.ChangeType],
-					NodeID:  uint64(crt.Replica.ReplicaID),
-					Context: encodedCtx,
-				})
+			return r.mu.raftGroup.ProposeConfChange(raftpb.ConfChange{
+				Type:    changeTypeInternalToRaft[crt.ChangeType],
+				NodeID:  uint64(crt.Replica.ReplicaID),
+				Context: encodedCtx,
+			})
 		}
 	}
 	return r.mu.raftGroup.Propose(encodeRaftCommand(string(p.idKey), data))

--- a/storage/replica_command.go
+++ b/storage/replica_command.go
@@ -28,6 +28,8 @@ import (
 	"math"
 	"time"
 
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
 	"golang.org/x/net/context"
 
 	"github.com/cockroachdb/cockroach/base"
@@ -42,7 +44,6 @@ import (
 	"github.com/cockroachdb/cockroach/util/protoutil"
 	"github.com/cockroachdb/cockroach/util/timeutil"
 	"github.com/cockroachdb/cockroach/util/uuid"
-	"github.com/coreos/etcd/raft"
 )
 
 var errTransactionUnsupported = errors.New("not supported within a transaction")
@@ -736,8 +737,8 @@ func (r *Replica) runCommitTrigger(ctx context.Context, batch engine.Batch, ms *
 			}
 			*ms = engine.MVCCStats{} // clear stats, as merge recomputed.
 		}
-		if ct.GetChangeReplicasTrigger() != nil {
-			if err := r.changeReplicasTrigger(ctx, batch, ct.ChangeReplicasTrigger); err != nil {
+		if crt := ct.GetChangeReplicasTrigger(); crt != nil {
+			if err := r.changeReplicasTrigger(ctx, batch, crt); err != nil {
 				return err
 			}
 		}
@@ -2506,50 +2507,88 @@ func (r *Replica) ChangeReplicas(
 	changeType roachpb.ReplicaChangeType, replica roachpb.ReplicaDescriptor, desc *roachpb.RangeDescriptor,
 ) error {
 
-	// Validate the request and prepare the new descriptor.
-	updatedDesc := *desc
-	updatedDesc.Replicas = append([]roachpb.ReplicaDescriptor(nil), desc.Replicas...)
-
-	found := -1       // tracks NodeID && StoreID
+	replicaIdx := -1  // tracks NodeID && StoreID
 	nodeUsed := false // tracks NodeID only
 	for i, existingRep := range desc.Replicas {
 		nodeUsed = nodeUsed || existingRep.NodeID == replica.NodeID
 
 		if existingRep.NodeID == replica.NodeID && existingRep.StoreID == replica.StoreID {
-			found = i
+			replicaIdx = i
 			replica.ReplicaID = existingRep.ReplicaID
 			break
 		}
 	}
+
+	rangeID := desc.RangeID
+	updatedDesc := *desc
+	updatedDesc.Replicas = append([]roachpb.ReplicaDescriptor(nil), desc.Replicas...)
 
 	switch changeType {
 	case roachpb.ADD_REPLICA:
 		// If the replica exists on the remote node, no matter in which store,
 		// abort the replica add.
 		if nodeUsed {
-			return util.Errorf("adding replica %v which is already present in range %d",
-				replica, desc.RangeID)
+			return util.Errorf("adding replica %v which is already present in range %d", replica, rangeID)
 		}
+
+		// Send a pre-emptive snapshot. Note that the replica to which this
+		// snapshot is addressed has not yet had its replica ID initialized; this
+		// is intentional, and serves to avoid the following race with the replica
+		// GC queue:
+		//
+		// - snapshot received, a replica is lazily created with the "real" replica ID
+		// - the replica is eligible for GC because it is not yet a member of the range
+		// - GC queue runs, creating a raft tombstone with the replica's ID
+		// - the replica is added to the range
+		// - lazy creation of the replica fails due to the raft tombstone
+		//
+		// Instead, the replica GC queue will create a tombstone with replica ID
+		// zero, which is never legitimately used, and thus never interferes with
+		// raft operations. Racing with the replica GC queue can still partially
+		// negate the benefits of pre-emptive snapshots, but that is a recoverable
+		// degradation, not a catastrophic failure.
+
+		snap, err := r.GetSnapshot()
+		if err != nil {
+			return util.Errorf("change replicas of %d failed: %s", rangeID, err)
+		}
+
+		fromReplica := *r.GetReplica()
+
+		if err := r.store.ctx.Transport.Send(&RaftMessageRequest{
+			GroupID:     r.RangeID,
+			FromReplica: fromReplica,
+			ToReplica:   replica,
+			Message: raftpb.Message{
+				Type:     raftpb.MsgSnap,
+				To:       uint64(replica.ReplicaID),
+				From:     uint64(fromReplica.ReplicaID),
+				Snapshot: snap,
+			},
+		}); err != nil {
+			return util.Errorf("change replicas of %d failed: %s", rangeID, err)
+		}
+
 		replica.ReplicaID = updatedDesc.NextReplicaID
 		updatedDesc.NextReplicaID++
 		updatedDesc.Replicas = append(updatedDesc.Replicas, replica)
 	case roachpb.REMOVE_REPLICA:
 		// If that exact node-store combination does not have the replica,
 		// abort the removal.
-		if found == -1 {
-			return util.Errorf("removing replica %v which is not present in range %d",
-				replica, desc.RangeID)
+		if replicaIdx == -1 {
+			return util.Errorf("removing replica %v which is not present in range %d", replica, rangeID)
 		}
-		updatedDesc.Replicas[found] = updatedDesc.Replicas[len(updatedDesc.Replicas)-1]
+		updatedDesc.Replicas[replicaIdx] = updatedDesc.Replicas[len(updatedDesc.Replicas)-1]
 		updatedDesc.Replicas = updatedDesc.Replicas[:len(updatedDesc.Replicas)-1]
 	}
 
+	descKey := keys.RangeDescriptorKey(desc.StartKey)
+
 	if pErr := r.store.DB().Txn(func(txn *client.Txn) error {
+		b := txn.NewBatch()
+
 		// Important: the range descriptor must be the first thing touched in the transaction
 		// so the transaction record is co-located with the range being modified.
-		b := &client.Batch{}
-		descKey := keys.RangeDescriptorKey(updatedDesc.StartKey)
-
 		if err := updateRangeDescriptor(b, descKey, desc, &updatedDesc); err != nil {
 			return err
 		}
@@ -2585,7 +2624,7 @@ func (r *Replica) ChangeReplicas(
 		})
 		return txn.Run(b)
 	}); pErr != nil {
-		return util.Errorf("change replicas of %d failed: %s", desc.RangeID, pErr)
+		return util.Errorf("change replicas of %d failed: %s", rangeID, pErr)
 	}
 	return nil
 }

--- a/storage/replica_raftstorage.go
+++ b/storage/replica_raftstorage.go
@@ -31,6 +31,7 @@ import (
 	"github.com/cockroachdb/cockroach/util"
 	"github.com/cockroachdb/cockroach/util/log"
 	"github.com/cockroachdb/cockroach/util/protoutil"
+	"github.com/cockroachdb/cockroach/util/retry"
 	"github.com/cockroachdb/cockroach/util/timeutil"
 )
 
@@ -426,6 +427,41 @@ func (r *Replica) Snapshot() (raftpb.Snapshot, error) {
 	return raftpb.Snapshot{}, raft.ErrSnapshotTemporarilyUnavailable
 }
 
+// GetSnapshot wraps Snapshot() but does not require the replica lock
+// to be held and it will block instead of returning
+// ErrSnapshotTemporaryUnavailable.
+func (r *Replica) GetSnapshot() (raftpb.Snapshot, error) {
+	retryOptions := retry.Options{
+		InitialBackoff: 1 * time.Millisecond,
+		MaxBackoff:     50 * time.Millisecond,
+		Multiplier:     2,
+	}
+	for retry := retry.Start(retryOptions); retry.Next(); {
+		r.mu.Lock()
+		snap, err := r.Snapshot()
+		snapshotChan := r.mu.snapshotChan
+		r.mu.Unlock()
+		if err == raft.ErrSnapshotTemporarilyUnavailable {
+			if snapshotChan == nil {
+				// The call to Snapshot() didn't start an async process due to
+				// rate limiting. Try again later.
+				continue
+			}
+			var ok bool
+			snap, ok = <-snapshotChan
+			if ok {
+				return snap, nil
+			}
+			// Each snapshot worker's output can only be consumed once.
+			// We could be racing with raft itself, so if we get a closed
+			// channel loop back and try again.
+		} else {
+			return snap, err
+		}
+	}
+	panic("unreachable") // due to infinite retries
+}
+
 func snapshot(
 	snap engine.Reader,
 	rangeID roachpb.RangeID,
@@ -602,15 +638,14 @@ func (r *Replica) applySnapshot(batch engine.Batch, snap raftpb.Snapshot) (uint6
 	// Extract the updated range descriptor.
 	desc := snapData.RangeDescriptor
 
-	replicaID := desc.NextReplicaID
-	if replicaDesc := r.GetReplica(); replicaDesc != nil {
-		replicaID = replicaDesc.ReplicaID
-	}
+	r.mu.Lock()
+	replicaID := r.mu.replicaID
+	r.mu.Unlock()
 
-	log.Infof("replica %s received snapshot for range %s at index %d. encoded size=%d, %d KV pairs, %d log entries",
+	log.Infof("replica %d received snapshot for range %d at index %d. encoded size=%d, %d KV pairs, %d log entries",
 		replicaID, rangeID, snap.Metadata.Index, len(snap.Data), len(snapData.KV), len(snapData.LogEntries))
 	defer func(start time.Time) {
-		log.Infof("replica %s applied snapshot for range %s in %s", replicaID, rangeID, timeutil.Since(start))
+		log.Infof("replica %d applied snapshot for range %d in %s", replicaID, rangeID, timeutil.Since(start))
 	}(timeutil.Now())
 
 	// Delete everything in the range and recreate it from the snapshot.
@@ -690,8 +725,10 @@ func (r *Replica) applySnapshot(batch engine.Batch, snap raftpb.Snapshot) (uint6
 	}
 
 	batch.Defer(func() {
-		// Update the range stats.
+		// Update the range and store stats.
+		r.store.metrics.subtractMVCCStats(r.stats.mvccStats)
 		r.stats.Replace(newStats)
+		r.store.metrics.addMVCCStats(r.stats.mvccStats)
 
 		r.mu.Lock()
 		// As outlined above, last and applied index are the same after applying

--- a/storage/store.go
+++ b/storage/store.go
@@ -1572,7 +1572,6 @@ func (s *Store) processRangeDescriptorUpdateLocked(rng *Replica) error {
 
 	// Add the range and its current stats into metrics.
 	s.metrics.replicaCount.Inc(1)
-	s.metrics.addMVCCStats(rng.stats.GetMVCC())
 
 	if s.mu.replicasByKey.Has(rng) {
 		return rangeAlreadyExists{rng}
@@ -1961,7 +1960,7 @@ func (s *Store) handleRaftMessage(req *RaftMessageRequest) error {
 		// or the past?
 		if !found && req.FromReplica.ReplicaID < desc.NextReplicaID {
 			if log.V(2) {
-				log.Infof("range %s: discarding message from replica %v, older than NextReplicaID %v",
+				log.Infof("range %s: discarding message from %+v, older than NextReplicaID %d",
 					req.GroupID, req.FromReplica, desc.NextReplicaID)
 			}
 			return nil
@@ -2000,7 +1999,7 @@ func (s *Store) handleRaftMessage(req *RaftMessageRequest) error {
 		// sender then we could prompt the sender to GC this replica
 		// immediately.
 		if log.V(1) {
-			log.Infof("refusing incoming Raft message %s for group %d from %s to %s: %s",
+			log.Infof("refusing incoming Raft message %s for group %d from %+v to %+v: %s",
 				req.Message.Type, req.GroupID, req.FromReplica, req.ToReplica, err)
 		}
 		return nil

